### PR TITLE
test(844): guard that workflows reference OIDC identifiers, never raw credentials

### PR DESCRIPTION
**Pull request by clarkemoyer**
Refs #844 — **acceptance criterion 4**, and a status correction on the rest of the issue.

> _"A test asserting no environment holds a secret outside the `*_AZURE_KV_*` allowlist, so the next raw credential pasted into GitHub fails CI instead of going unnoticed for months."_

## The finding: phases 1 and 2 are already code-complete

I picked this issue expecting to convert workflows. There is nothing left to convert. The tracking comment on #844 says **"13 workflows still on `CBM_TOKEN`: 120, 401, 502, 701, 702, 703, 704, 720, 729, 735, 736, 737, 742"**. The real count, across the whole tree:

```
$ grep -rn "secrets.CBM_TOKEN" --include=*.yml --include=*.ps1 --include=*.py .
(nothing)
```

The only surviving hit anywhere is one stale sentence in `README.md`. Every `CBM_TOKEN` string still in a workflow is a **comment** describing auth that no longer happens. Phase 1 (`whmcs-prod`) is likewise done — `WHMCS_API_ACCESS_KEY` and the WHMCS credential pair appear nowhere.

That gap between the tracking prose and the tree is the reason this PR is a test rather than a doc edit. A list of remaining work maintained by hand describes the tree at the moment someone last looked at it. The next session to pick up #844 would have re-derived exactly what I did, or — worse — trusted the comment and "converted" workflows that were already converted.

**Complete current picture**, every `${{ secrets.X }}` in all 97 workflows:

| Secret | Kind | Status |
| --- | --- | --- |
| `GITHUB_TOKEN` | ambient per-run token | not a stored credential |
| `*_AZURE_CLIENT_ID`, `*_AZURE_KV_CLIENT_ID`, `*_AZURE_TENANT_ID` | Entra app + tenant GUIDs | identifiers, authorize nothing |
| `FFC_WPMUDEV_GA_API_Token` | credential | **phase 3** — 104, 601 |
| `FFC_EXO_CERT_PFX_BASE64`, `FFC_EXO_CERT_PASSWORD` | credential | **phase 4** — 103, 304 |
| `EXO_CERT_PFX_BASE64`, `EXO_CERT_PFX_PASSWORD` | credential | **phase 4** — 303 |
| `EXO_TENANT`, `EXO_ORGANIZATION` | configuration | **phase 4** — 303 |

## What the test can and cannot assert

It cannot enumerate the secrets **stored** in an environment — that needs an authenticated admin-scoped call, and giving CI that scope to check for over-scoped credentials would be its own problem.

But a stored secret is **inert until a workflow references it**, and the reference is in the tree. So the reference is what gets pinned. A credential pasted into GitHub is only dangerous once something reads it, and that read now has to pass review against an allowlist.

The two `EXO_TENANT` / `EXO_ORGANIZATION` entries are configuration, not credentials — a tenant domain and an org name. They are enumerated anyway, because the criterion is that a migrated environment holds *only* `*_AZURE_KV_*` identifiers, and these are what keep `m365-prod` from being clean. They need no Key Vault at all; repo Variables or workflow inputs would do.

## Four guards, each mutation-tested

| Test | Fails when |
| --- | --- |
| `test_workflows_reference_no_unexpected_secrets` | a new raw credential is referenced — the headline criterion |
| `test_the_remaining_raw_credentials_list_is_exact` | debt is migrated but its line is left behind |
| `test_migrated_credentials_are_not_reintroduced` | `CBM_TOKEN` / `WHMCS_API_*` comes back as a GitHub secret |
| `test_key_vault_consumers_can_perform_the_oidc_exchange` | a KV-fetching **job** lacks `id-token: write` |

**The debt list is exact, not a ceiling.** Both directions matter, and the second is the one that would otherwise rot: adding a credential fails, and so does *removing* one without deleting its entry. The list cannot drift into a stale overstatement of the work the way the tracking comment did. (Same reasoning as #837's enumeration guard, and the #822 grooming note: a count is not evidence, the list is.)

**The fourth guard is about the failure mode, not the mechanism.** A job that fetches from Key Vault without `id-token: write` fails at run time with an opaque `azure/login` error. The tempting fix under time pressure is to paste the credential back into a GitHub secret — the exact regression the other three guards exist to prevent, arrived at through a route they would not have caught in time.

## Two review rounds, and the one defect behind all of them

Copilot filed four findings across two rounds. **Every one was correct**, and every one was the same defect wearing a different hat: the guard recognizing **one spelling of a thing that has several**, and passing green over the rest.

| Round | Finding | What it missed |
| --- | --- | --- |
| 1 | consumer detection required the literal `az keyvault secret show` in the workflow | **56 of 71** consumers reach Key Vault via a `*-from-kv` composite action and never contain the string |
| 1 | `permissions:` checked file-wide | a job-level block **replaces** the workflow-level one, so a two-job workflow could pass and still fail at run time |
| 2 | `_SECRET_REF` matched only `secrets.NAME` | `secrets['NAME']` is exactly equivalent — a one-character bypass of the whole allowlist |
| 2 | detection pinned the `show` subcommand | the repo also uses `secret download` (5 sites) and `secret list` (5 sites); 4 jobs were genuinely uncovered |

The first is the one worth dwelling on: guard 4 originally **passed while inspecting 15 of 71 workflows**. A test reading green while checking almost nothing is precisely the failure class the rest of this file is written against, and it would have held for as long as nobody looked.

The fixes generalize rather than patch. Consumer detection resolves `uses: ./.github/actions/<x>` against the set of local composite actions that themselves call Key Vault — **derived from the tree, not listed** — so a new KV-fetching action is covered the day it lands and a renamed one cannot silently drop out. That set is asserted non-empty, so a layout move fails loudly instead of quietly finding nothing. Permissions are evaluated per job. Detection matches any `az keyvault` call. The reference regex matches both syntaxes.

The shape is now recorded in the module docstring, because the next check added to this file will be tempted into the same mistake.

**Coverage: 15 → 75 workflows, 93 Key Vault-consuming jobs.** All already declare `id-token: write`, so none of this adds a failure — it adds the ability to notice one.

## Mutation results

Each mutation fires its intended test and only that test; each recovers on restore. The four review-round cases are shown against the form they slipped, confirming the findings were real rather than theoretical.

| Mutation | Old form | Now |
| --- | --- | --- |
| `secrets.GITHUB_TOKEN` → `secrets.SOME_NEW_API_TOKEN` in 725 | — | `…no_unexpected_secrets` |
| add `GH_TOKEN: ${{ secrets.CBM_TOKEN }}` to 737 | — | `…not_reintroduced` **+** `…no_unexpected_secrets` (with the KV-secret-name message, not the generic one) |
| migrate 104's wpmudev reference, leave the list untouched | — | `…list_is_exact` |
| drop `id-token: write` from 726 (**inline** consumer) | FAIL | FAIL (unchanged) |
| drop `id-token: write` from 202 (**composite-action** consumer) | **PASS** ❌ | **FAIL** ✅ |
| job-level `permissions:` block omitting `id-token` on 202 | **PASS** ❌ | **FAIL** ✅ |
| raw credential via `secrets['SMUGGLED_API_KEY']` | **not seen at all** ❌ | **FAIL** ✅ |
| drop `id-token: write` from 505 (**`download`-only** consumer) | **not a consumer** ❌ | **FAIL** ✅ |

## Verification

Re-run on the merged tree after syncing `main` (9 commits, the 506 GA4 fixes + a CI change) — which matters here, because an exact-list guard turns red if a workflow arriving on `main` adds a secret reference. It didn't.

- `tests/workflow-logic/run_all.py` — **27/27 modules pass**. The harness runs cleanly in this sandbox; no false-red to discount.
- `prettier@3.8.1 --check .` (CI-pinned, full repo) ✅ · `check-workflow-doc-consistency.py` ✅ 97/90 · `check-workflow-references.py` ✅ · catalog regenerates with **no drift**.
- `actionlint` not run — **no workflow file is touched by this PR**. One new test module, nothing else.

## Deliberately not in this PR

**The stale comments.** Eight workflows plus `README.md` still describe `CBM_TOKEN` auth that no longer happens. Fixing them is right, but the catalog embeds each workflow's header comment, so touching seven headers regenerates seven catalog entries — and #837, #858 and #878 are all open and all regenerate that file. A near-certain conflict for a prose fix. Better as its own PR once those land; noted on #844 so it is not lost.

**`test_env_scoped_secrets.py`'s `CBM_TOKEN → {github-prod}` entry**, which is now a no-op. It stays: it correctly guards the reintroduction case, it costs nothing, and #837 modifies that file.

## What still needs a human

**The GitHub secrets themselves are still there.** Code-complete is not done — nothing in this repo can delete an environment secret, and @clarkemoyer should now be able to remove `CBM_TOKEN` from `github-prod` and the two legacy WHMCS secrets from `whmcs-prod`, since nothing reads them. Until that happens the credential still exists in two places and can still drift; this PR only guarantees no *workflow* will read the GitHub copy.

Phases 3 and 4 need Azure work an agent cannot do: one federated credential for `…:environment:wpmudev-prod` (the KV secret already exists), and the EXO `.pfx` uploaded to Key Vault by a human — #844 is explicit that no agent should handle that material.

## Gates

None. No `environment:`, no credentials, no workflow changed — one new test module. Normal merge queue.

### CODE REVIEW

File: tests/workflow-logic/test_no_raw_credential_secrets.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    `az keyvault secret show` without the permission gets an opaque azure/login
+    failure — and the tempting fix, under time pressure, is to paste the
+    credential back into a GitHub secret, which is what this file exists to stop.
+    """
+    violations = []
+    for path in sorted(WORKFLOWS.glob("*.yml")):
+        text = path.read_text(encoding="utf-8")
+        if "az keyvault secret show" not in text:
+            continue
+        if "id-token: write" not in text:
```
[the remark this answers is not in the document]
Comment (clarkemoyer) [reply tests/workflow-logic/test_no_raw_credential_secrets.py]: Correct, and worse than stated. Fixed in `5bc9ed8`.

The substring form matched **15** workflows. The other **56** reach Key Vault only through a `*-from-kv` composite action and never contain the string — so the guard reported on 71 workflows while inspecting 15. Not a missed edge case: a test reading green while checking almost nothing, which is the exact failure class the rest of this file is written against. It would have passed for as long as nobody looked.

**Detection** now resolves `uses: ./.github/actions/<x>` against the set of local composite actions that themselves call `az keyvault secret show`. Derived from the tree rather than listed — a new KV-fetching action is covered the day it lands, and matching on the call rather than the `-from-kv` name means a differently-named or renamed action cannot silently drop out of coverage. Currently resolves to 6 actions (`candid-keys`, `cloudflare-tokens`, `fraudlabspro-keys`, `google-secrets`, `whmcs-secrets`, `zeffy-secrets`), asserted non-empty so a layout move fails loudly instead of quietly finding nothing.

**Your parenthetical `(job-level)` turned out to be a second bug**, so thank you for it. The old check was file-wide, but `permissions:` is per job and a job-level block *replaces* the workflow-level one rather than merging. A two-job workflow whose non-KV job declares `id-token: write` passed the substring test and would still have failed at run time. Now evaluated per job, workflow-level used only as the fallback.

Both mutation-tested, and both verified to **pass under the old form** — confirming the finding was real rather than theoretical:

| Mutation | Old | New |
| --- | --- | --- |
| drop `id-token: write` from 202 (composite-action consumer) | PASS | **FAIL** |
| add a job-level `permissions:` block omitting `id-token` to 202 | PASS | **FAIL** |
| drop `id-token: write` from 726 (inline consumer) | FAIL | FAIL (unchanged) |

27/27 workflow-logic modules pass; prettier, doc-consistency, reference guard and catalog clean. No workflow file touched.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/workflow-logic/test_no_raw_credential_secrets.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+import yaml
+
+sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent))
+from wf_extract import REPO_ROOT, WORKFLOWS
+
+# Only matches inside a ${{ ... }} expression, so prose ("…/environments/…/secrets")
+# and PowerShell property access ($secrets.Count) are not mistaken for references.
+_EXPRESSION = re.compile(r"\$\{\{(.*?)\}\}", re.DOTALL)
+_SECRET_REF = re.compile(r"\bsecrets\.([A-Za-z_][A-Za-z0-9_]*)")
+
```
[the remark this answers is not in the document]
Comment (clarkemoyer) [reply tests/workflow-logic/test_no_raw_credential_secrets.py:74]: Right, and it bypassed the allowlist completely rather than partially. Fixed in `2e23694`, along with both suppressed-confidence findings from the same review — all three were correct.

**Bracket notation.** `_SECRET_REF` now matches `secrets.NAME`, `secrets['NAME']` and `secrets["NAME"]`. Nothing in the tree uses the bracket form, which is precisely why it needed covering: a guard is worth what it catches in the syntax nobody has written yet, and one character was the entire cost of evading this one.

**The two suppressed comments were also right, and I'd rather they hadn't been suppressed.** Detection pinned `az keyvault secret show`, but this repo also uses `secret download` (5 call sites) and `secret list` (5 call sites). Every one needs the same OIDC-issued Azure session, so pinning a subcommand dropped the rest silently. Now matches `az keyvault`.

Your call-out of **501 was specific enough to verify, and it checks out** — `az keyvault secret download` at line 109. 501 stayed covered only because it *also* uses a composite action, i.e. incidentally rather than by design. Four jobs were genuinely uncovered:

`321:monitor` · `503:provision` · `504:backup` · `505:provision`

Coverage: **71 → 75 workflows, 93 KV-consuming jobs.** All four already declare `id-token: write`, so this adds no failures — it adds the ability to notice.

Both mutation-tested, and both confirmed to slip the previous form:

| Mutation | Old | New |
| --- | --- | --- |
| raw credential via `secrets['SMUGGLED_API_KEY']` | not seen at all | **FAIL** |
| drop `id-token: write` from 505 (`download`-only consumer) | not treated as a consumer | **FAIL** |

Worth noting the shape across both rounds: every finding has been the same defect wearing a different hat — the guard recognizing one spelling of a thing that has several. Dot vs bracket, `show` vs `download`/`list`, inline vs composite action, file-wide vs per-job. Each pinned a form rather than the thing itself. That is now the standing note in the module header for whoever extends it.

27/27 workflow-logic modules pass; prettier, doc-consistency, reference guard and catalog clean. No workflow file touched.

---
_Generated by [Claude Code](https://claude.ai/code)_
